packages/world/src/utils/PlayerUtils.sol

```diff
@@ -30,110 +31,132 @@ import { MAX_PLAYER_INFLUENCE_HALF_WIDTH, PLAYER_ENERGY_DRAIN_RATE, FRAGMENT_SIZ
 
 using ObjectTypeLib for ObjectTypeId;
 
-function requireValidPlayer(address player) returns (EntityId, Vec3, EnergyData memory) {
-  checkWorldStatus();
-  EntityId playerEntityId = Player._get(player);
-  require(playerEntityId.exists(), "Player does not exist");
-  require(!PlayerStatus._getBedEntityId(playerEntityId).exists(), "Player is sleeping");
-  Vec3 playerCoord = PlayerPosition._get(playerEntityId);
+library PlayerUtils {
+  // TODO: currently a public lib function due to contract size limit, should ideally figure out a better workaround
+  function requireValidPlayer(address player) public returns (EntityId, Vec3, EnergyData memory) {
+    checkWorldStatus();
+    EntityId playerEntityId = Player._get(player);
+    require(playerEntityId.exists(), "Player does not exist");
+    require(!PlayerStatus._getBedEntityId(playerEntityId).exists(), "Player is sleeping");
+    Vec3 playerCoord = PlayerPosition._get(playerEntityId);
 
-  EnergyData memory playerEnergyData = updateEnergyLevel(playerEntityId);
-  require(playerEnergyData.energy > 0, "Player is dead");
+    EnergyData memory playerEnergyData = updatePlayerEnergy(playerEntityId);
+    require(playerEnergyData.energy > 0, "Player is dead");
```

Comment:
so the idea is if you're dead, then all txs like build/mine/move/drop etc will revert, and so you need to call spawn

in the client, we'll know if they're dead, so no need to call activate, just show death/respawn screen

we'll still use activate to kill players who never load the client

Reply:
but also this means if you mine/build and lose energy, it doesnt kill you right?

Reply:
do you think its better if it kills you ie remove from grid? or better if client shows just shows death and only removes you once you call spawn/activate

Reply:
It kills you in the sense that it can leave you with 0 energy, but yeah you need to respawn or call activate. We could kill whenever energy gets to 0, but it was a little bit more complex

Reply:
> do you think its better if it kills you ie remove from grid? or better if client shows just shows death and only removes you once you call spawn/activate

Not sure tbh, if we are worried about consistency then maybe we can just kill whenever energy gets to 0

Reply:
I think its fine for build/mine but I'm worried about moves, and later hits. If you mine a block and drop someone, you'd expect them to be removed from the grid. Same with hitting. Otherwise the client would have to call `activate` on these players

Reply:
Ah we also need to take into account gas costs, having too many items could make you prevent certain deaths if we remove on death

Reply:
true, this will be true for hits too. In these cases though, I assume the attacker is willing to pay the gas cuz you want the other persons items. As long as the tx fits within block limit haha. This code path should also be rare